### ISSUE TITLE
[Bug] tf.contrib.layers.layer_norm: third-party implementation does not reflect the original paper

### ISSUE THREAD

**Issue by xcyan** [tensorflow/tensorflow#14562]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.3
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: CUDA 8.0/cuDNN 6.0
- **GPU model and memory**: Titan X
- **Exact command to reproduce**: 

### Describe the problem
In the [original layer normalization paper](https://arxiv.org/pdf/1607.06450.pdf), it's been said that beta/gamma should have the same dimension as hidden variable (see the lines below Eq (4) in Page 3). Based on current implementation, there is no such option that reflects the original paper.

Please run the following source code to verify that.

Also, see the comment #3671 .

### Source code / logs
`import tensorflow as tf`
`layer_norm = tf.contrib.layers.layer_norm`
`batch_size = 10`
`hidden_dim = 5`
`input = tf.zeros((batch_size, hidden_dim), dtype=tf.float32)`
`output = layer_norm(input)`
`tf.trainable_variables()`

**Comment by rohan100jain** [comment]
@xodus7 Could you take a look at this please?

**Comment by xcyan** [comment]
@rohan100jain Is it possible to assign this to TF developers? I doubt the original author will take a look due to his inactive github status.

**Comment by xodus7** [comment]
Hi sorry, but I've been traveling these past few days and haven't had time to look at this properly until now. In your example code above alpha and gamma have a shape of (5,) which is consistent with what's expected. What results are you getting?

**Comment by xcyan** [comment]
Shouldn't the size be (5, 10)? If you refer to the sentences in this paper.

> b and g are defined as the bias
and gain parameters of the same dimension as h_t

**Comment by xodus7** [comment]
No, that equation does not include the batch or time dimensions. See the last few sentences of the first paragraph in section 3.1. This is one of the big advantages of layer normalization over batch norm for RNNs.

**Comment by xcyan** [comment]
Please compare (3) and (4): mu and sigma are scalars while beta and gamma
are vectors.

This means beta and gamma should have one more rank compared to mu and
sigma.

If you consider batch_size as well, mu and sigma should have size
(batch_size) while beta and gamma should have size (batch_size, gamma).

**Comment by xodus7** [comment]
In equation (4): a^t = W_hh*a^(t-1) + W_xh*x^t yeilding a rank 1 vector of size h. This h_t is also a rank one vector of size h. The batch and time dimensions are not considered here and Gamma and beta are shared (broadcasted) across them.

If gamma was (batch_size, hidden_size) then we'd be learning a separate scaling factor per batch per channel which does not match up with the paper's description.

**Comment by xcyan** [comment]
@xodus7 I see your point. Is it possible to add some unit test for the layer_normalization.

**Comment by xodus7** [comment]
@xcyan There are some tests: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers_test.py#L2750

What coverage do you feel is missing?

**Comment by xcyan** [comment]
What we discussed: the shape of gamma and beta.

**Comment by xodus7** [comment]
Not a bad idea; if I get some free time I'll add them. Please feel free to add some yourself and/or let me know if you find any problems. Are you using this op in your projects?

**Comment by xcyan** [comment]
@xodus7 Yes. Let us chat offline since it's not relevant to TensorFlow issue.
